src/neo4j/mod.rs

```diff
@@ -0,0 +1,702 @@
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Neo4j {
+    version: Value,
+    user: Value,
+    pass: Value,
+    enterprise: bool,
+    plugins: Vec<Neo4jLabsPlugin>,
+}
+
+impl Neo4j {
+    const DEFAULT_USER: &'static str = "neo4j";
+    const DEFAULT_PASS: &'static str = "neo";
+    const DEFAULT_VERSION_TAG: &'static str = "5";
+
+    /// Create a new instance of a Neo4j image.
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            version: Value::Default(Self::DEFAULT_VERSION_TAG),
+            user: Value::Default(Self::DEFAULT_USER),
+            pass: Value::Default(Self::DEFAULT_PASS),
+            enterprise: false,
+            plugins: Vec::new(),
+        }
+    }
+
+    /// Create a new instance of a Neo4j 5 image with the default user and password.
+    #[must_use]
+    pub const fn from_env() -> Self {
+        Self {
+            version: Value::Env {
+                var: "NEO4J_VERSION_TAG",
+                fallback: Self::DEFAULT_VERSION_TAG,
+            },
+            user: Value::Env {
+                var: "NEO4J_TEST_USER",
+                fallback: Self::DEFAULT_USER,
+            },
+            pass: Value::Env {
+                var: "NEO4J_TEST_PASS",
+                fallback: Self::DEFAULT_PASS,
+            },
+            enterprise: false,
+            plugins: Vec::new(),
+        }
+    }
+
+    /// Set the Neo4j version to use.
+    /// The value must be an existing Neo4j version tag.
+    ///
+    /// Only a subset of Semantic Versions are supported.
+    /// The version must be of the format
+    ///
+    ///    MAJOR[.MINOR[.PATCH]]
+    ///
+    ///
+    /// # Errors
+    ///
+    /// If the version is not valid according to the format.
+    pub fn with_version(
+        mut self,
+        version: impl Into<String>,
+    ) -> Result<Self, Box<dyn std::error::Error + Sync + Send + 'static>> {
+        let version: String = version.into();
+
+        let version_valid =
+            lenient_semver::parse_into::<'_, ValidateVersion>(&version).unwrap_or(false);
+        if !version_valid {
+            return Err(format!("Invalid version: {}", version).into());
+        }
+
+        self.version = Value::Value(version);
+        Ok(self)
+    }
+
+    /// Set the username to use.
+    #[must_use]
+    pub fn with_user(mut self, user: impl Into<String>) -> Self {
+        self.user = Value::Value(user.into());
+        self
+    }
+
+    /// Set the password to use.
+    #[must_use]
+    pub fn with_password(mut self, pass: impl Into<String>) -> Self {
+        self.pass = Value::Value(pass.into());
+        self
+    }
+
+    /// Do not use any authentication on the testcontainer.
+    ///
+    /// Setting this will override any prior usages of [`Self::with_user`] and
+    /// [`Self::with_password`].
+    pub fn without_authentication(mut self) -> Self {
+        self.user = Value::Unset;
```

Comment:
This could call with_user and with_password internally if I'm not mistaken.

Would make it easier to maintain IMO.

Reply:
Whoops. Missed that.